src/autopas/selectors/AutoTuner.h

```diff
@@ -457,74 +459,87 @@ bool AutoTuner<Particle>::iteratePairwise(PairwiseFunctor *f, bool doListRebuild
+    // 1. particleBuffer with all close particles in container
+    for (auto &&p : particleBuffer) {
+      auto pos = p.getR();
+      auto min = autopas::utils::ArrayMath::subScalar(pos, cutoff);
+      auto max = autopas::utils::ArrayMath::addScalar(pos, cutoff);
+      for (auto iter2 = containerPtr->getRegionIterator(min, max, IteratorBehavior::ownedOrHalo); iter2.isValid();
+           ++iter2) {
+        if (newton3) {
+          f->AoSFunctor(p, *iter2, true);
+        } else {
+          f->AoSFunctor(p, *iter2, false);
+          f->AoSFunctor(*iter2, p, false);
+        }
       }
     }
-  }
 
-  // 2. haloParticleBuffer with owned, close particles in container
-  for (auto &&p : haloParticleBuffer) {
-    auto pos = p.getR();
-    auto min = autopas::utils::ArrayMath::subScalar(pos, containerPtr->getCutoff());
-    auto max = autopas::utils::ArrayMath::addScalar(pos, containerPtr->getCutoff());
-    for (auto iter2 = containerPtr->getRegionIterator(min, max, IteratorBehavior::owned); iter2.isValid(); ++iter2) {
-      if (newton3) {
-        f->AoSFunctor(p, *iter2, true);
-      } else {
-        // Here, we do not need to interact p with *iter2, because p is a halo particle and an AoSFunctor call with an
-        // halo particle as first argument has no effect if newton3 == false.
-        f->AoSFunctor(*iter2, p, false);
+    // 2. haloParticleBuffer with owned, close particles in container
+    for (auto &&p : haloParticleBuffer) {
+      auto pos = p.getR();
+      auto min = autopas::utils::ArrayMath::subScalar(pos, cutoff);
+      auto max = autopas::utils::ArrayMath::addScalar(pos, cutoff);
+      for (auto iter2 = containerPtr->getRegionIterator(min, max, IteratorBehavior::owned); iter2.isValid(); ++iter2) {
+        if (newton3) {
+          f->AoSFunctor(p, *iter2, true);
+        } else {
+          // Here, we do not need to interact p with *iter2, because p is a halo particle and an AoSFunctor call with an
+          // halo particle as first argument has no effect if newton3 == false.
+          f->AoSFunctor(*iter2, p, false);
+        }
       }
     }
   }
 
-  // 3. particleBuffer with itself
-  for (size_t i = 0; i < particleBuffer.size(); ++i) {
-    auto &&p1 = particleBuffer[i];
-    for (size_t j = i + 1; j < particleBuffer.size(); ++j) {
-      if (i == j) {
-        continue;
-      }
-      auto &&p2 = particleBuffer[j];
-      if (newton3) {
-        f->AoSFunctor(p1, p2, true);
-      } else {
-        f->AoSFunctor(p1, p2, false);
-        f->AoSFunctor(p2, p1, false);
+  // 3. particleBuffer with itself and all other buffers
+  for (size_t bufferIdOuter = 0; bufferIdOuter < particleBuffers.size(); ++bufferIdOuter) {
+    auto &particleBufferOuter = particleBuffers[bufferIdOuter];
+    for (size_t i = 0; i < particleBufferOuter.size(); ++i) {
+      auto &&p1 = particleBufferOuter[i];
+      for (size_t bufferIdInner = bufferIdOuter; bufferIdInner < particleBuffers.size(); ++bufferIdInner) {
```

Comment:
Any reason why this loop is inside the "i" loop? Initial looks make it seem like these are interchangeable and the switch might help readability.

Reply:
This ordering reduces the number of memory accesses. We access each outer particle (`i`) once and then the whole inner buffer. If you change the loops you access each `i` once for every buffer. If the buffers are small this should not be significant but I feel they are sometimes not small...